### ISSUE TITLE
Postgrest 8 fails to start if standard_conforming_strings=off

### ISSUE THREAD

**Issue by dgtrapeze** [PostgREST/postgrest#1992]
### Environment

* PostgreSQL version: 9.6 (although tried postgresql 12 for the same outcome)
* PostgREST version: 8.0.0
* Operating system: Tried on RHEL8 and Centos6

### Description of issue
I just download postgrest 8.0.0 to try it out.  We previously used postgrest 7.0.1.  However, on starting up postgrest it just continually output

```
27/Oct/2021:08:43:40 +1000: Attempting to connect to the database...
27/Oct/2021:08:43:40 +1000: Connection successful
27/Oct/2021:08:43:40 +1000: Listening on port 3000
27/Oct/2021:08:43:40 +1000: Config re-loaded
27/Oct/2021:08:43:40 +1000: Listening for notifications on the pgrst channel
27/Oct/2021:08:43:40 +1000: An error ocurred when loading the schema cache
27/Oct/2021:08:43:40 +1000: {"hint":null,"details":"Token \"QUERY\" is invalid.","code":"22P02","message":"invalid input syntax for type json"}
```

<details>
<summary>The pg logs show the query being executed as the pfkSourceColumns query.</summary>

```
2021-10-27 08:43:59 AEST [2076]: [13-1] db=testdb,user=web_authenticator,app=Postgrest pool connection,client=[local] LOG:  execute 4:
              with recursive
              pks_fks as (
                -- pk + fk referencing col
                select
                  conrelid as resorigtbl,
                  unnest(conkey) as resorigcol
                from pg_constraint
                where contype IN ('p', 'f')
                union
                -- fk referenced col
                select
                  confrelid,
                  unnest(confkey)
                from pg_constraint
                where contype='f'
              ),
              views as (
                select
                  c.oid       as view_id,
                  n.nspname   as view_schema,
                  c.relname   as view_name,
                  r.ev_action as view_definition
                from pg_class c
                join pg_namespace n on n.oid = c.relnamespace
                join pg_rewrite r on r.ev_class = c.oid
                where c.relkind in ('v', 'm') and n.nspname = ANY($1 || $2)
              ),
              transform_json as (
                select
                  view_id, view_schema, view_name,
                  -- the following formatting is without indentation on purpose
                  -- to allow simple diffs, with less whitespace noise
                  replace(
                    replace(
                    replace(
                    replace(
                    replace(
                    replace(
                    replace(
                    replace(
                    regexp_replace(
                    replace(
                    replace(
                    replace(
                    replace(
                    replace(
                    replace(
                    replace(
                    replace(
                    replace(
                    replace(
                    replace(
                    replace(
                      view_definition::text,
                    -- This conversion to json is heavily optimized for performance.
                    -- The general idea is to use as few regexp_replace() calls as possible.
                    -- Simple replace() is a lot faster, so we jump through some hoops
                    -- to be able to use regexp_replace() only once.
                    -- This has been tested against a huge schema with 250+ different views.
                    -- The unit tests do NOT reflect all possible inputs. Be careful when changing this!
                    -- -----------------------------------------------
                    -- pattern           | replacement         | flags
                    -- -----------------------------------------------
                    -- `,` is not part of the pg_node_tree format, but used in the regex.
                    -- This removes all `,` that might be part of column names.
                       ','               , ''
                    -- The same applies for `{` and `}`, although those are used a lot in pg_node_tree.
                    -- We remove the escaped ones, which might be part of column names again.
                    ), '\{'              , ''
                    ), '\}'              , ''
                    -- The fields we need are formatted as json manually to protect them from the regex.
                    ), ' :targetList '   , ',"targetList":'
                    ), ' :resno '        , ',"resno":'
                    ), ' :resorigtbl '   , ',"resorigtbl":'
                    ), ' :resorigcol '   , ',"resorigcol":'
                    -- Make the regex also match the node type, e.g. `{QUERY ...`, to remove it in one pass.
                    ), '{'               , '{ :'
                    -- Protect node lists, which start with `({` or `((` from the greedy regex.
                    -- The extra `{` is removed again later.
                    ), '(('              , '{(('
                    ), '({'              , '{({'
                    -- This regex removes all unused fields to avoid the need to format all of them correctly.
                    -- This leads to a smaller json result as well.
                    -- Removal stops at `,` for used fields (see above) and `}` for the end of the current node.
                    -- Nesting can't be parsed correctly with a regex, so we stop at `{` as well and
                    -- add an empty key for the followig node.
                    ), ' :[^}{,]+'       , ',"":'              , 'g'
                    -- For performance, the regex also added those empty keys when hitting a `,` or `}`.
                    -- Those are removed next.
                    ), ',"":}'           , '}'
                    ), ',"":,'           , ','
                    -- This reverses the "node list protection" from above.
                    ), '{('              , '('
                    -- Every key above has been added with a `,` so far. The first key in an object doesn't need it.
                    ), '{,'              , '{'
                    -- pg_node_tree has `()` around lists, but JSON uses `[]`
                    ), '('               , '['
                    ), ')'               , ']'
                    -- pg_node_tree has ` ` between list items, but JSON uses `,`
                    ), ' '             , ','
                    -- `<>` in pg_node_tree is the same as `null` in JSON, but due to very poor performance of json_typeof
                    -- we need to make this an empty array here to prevent json_array_elements from throwing an error
                    -- when the targetList is null.
                    ), '<>'              , '[]'
                  )::json as view_definition
                from views
              ),
              target_entries as(
              ),
              target_entries as(
                select
                  view_id, view_schema, view_name,
                  json_array_elements(view_definition->0->'targetList') as entry
                from transform_json
              ),
              results as(
                select
                  view_id, view_schema, view_name,
                  (entry->>'resno')::int as view_column,
                  (entry->>'resorigtbl')::oid as resorigtbl,
                  (entry->>'resorigcol')::int as resorigcol
                from target_entries
              ),
              recursion as(
                select r.*
                from results r
                where view_schema = ANY ($1)
                union all
                select
                  view.view_id,
                  view.view_schema,
                  view.view_name,
                  view.view_column,
                  tab.resorigtbl,
                  tab.resorigcol
                from recursion view
                join results tab on view.resorigtbl=tab.view_id and view.resorigcol=tab.view_column
              )
              select
                sch.nspname as table_schema,
                tbl.relname as table_name,
                col.attname as table_column_name,
                rec.view_schema,
                rec.view_name,
                vcol.attname as view_column_name
              from recursion rec
              join pg_class tbl on tbl.oid = rec.resorigtbl
              join pg_attribute col on col.attrelid = tbl.oid and col.attnum = rec.resorigcol
              join pg_attribute vcol on vcol.attrelid = rec.view_id and vcol.attnum = rec.view_column
              join pg_namespace sch on sch.oid = tbl.relnamespace
              join pks_fks using (resorigtbl, resorigcol)
              order by view_schema, view_name, view_column_name;
2021-10-27 08:43:59 AEST [2076]: [14-1] db=testdb,user=web_authenticator,app=Postgrest pool connection,client=[local] DETAIL:  parameters: $1 = '{webservices}', $2 = '{bacchus,public}'
2021-10-27 08:43:59 AEST [2076]: [15-1] db=testdb,user=web_authenticator,app=Postgrest pool connection,client=[local] ERROR:  invalid input syntax for type json
2021-10-27 08:43:59 AEST [2076]: [16-1] db=testdb,user=web_authenticator,app=Postgrest pool connection,client=[local] DETAIL:  Token "QUERY" is invalid.
2021-10-27 08:43:59 AEST [2076]: [17-1] db=testdb,user=web_authenticator,app=Postgrest pool connection,client=[local] CONTEXT:  JSON data, line 1: [QUERY...
```
</details>

Our application started life on postgresql 7 and has upgraded along the way.  However, as it predates `standard_conforming_strings`, we need to run `standard_conforming_strings=off` for our application to run as lots of queries assume the old escape system.  This is set in postgresql.conf.

The above query assumes `standard_conforming_strings=on` and since we have it off, the query fails to work as expected and the resulting string which is cast to json is not correctly formatted.

To get postgrest to work again, I've had to do

```sql
alter role web_authenticator set standard_conforming_strings = on;
```

where web_authenticator is our role that postgrest uses to connect.

I suspect postgrest assumes standard conforming strings in other places as well now (eg escaping in query parameters was added in v8 I saw?).  

I'm not yet fully sure of the impacts to our application having set this for the postgrest authenticator role.  We are usually pretty good at using E'' strings now, but you never know what developers do....

I would assume the fix is for postgrest to use escape string syntax (E'' strings) to insulate it from these sorts of database settings.

It could automatically set `standard_conforming_strings=on` for its session (like it does for search_path) or you could add docs to do the alter role authenticator I had to do, but I suspect those options have side-effects.

(Expected behavior vs actual behavior)
I expected postgrest 8 to pretty much work the same as postgrest 7.

(Steps to reproduce: Include a minimal SQL definition plus how you make the request to PostgREST and the response body)

```sql
alter system set standard_conforming_strings = off;
```
then start postgrest

**Comment by rinshadka** [comment]
I have tried `v8.0.0.20211102 `and `v9.0.0-a1` pre-releases. But faced this same error although its mentioned that the `v8.0.0.20211102` is having the fix for this particular issue. 

Please find the logs below while starting the service, 


![2021-11-12 14_57_44-Window](https://user-images.githubusercontent.com/40381766/141721199-d7d3657d-a7c5-4970-a321-6cb9038ad10b.png)

Also experimentally changed the standard_conforming_strings postgreSQL configuration for postgREST database user, but failed.

`alter role postgrest_authenticator set standard_conforming_strings = on;`

Whenever I reverts the version to `v7.0.1` , everything works fine.

Thanks.

**Comment by steve-chavez** [comment]
@rinshadka Can you confirm that error happens when running the below query?

(Replace `$1 || $2` and `$1` with `ARRAY['your_schema']`)

<details>
<summary>pfkSourceColumns query</summary>
<pre>
with recursive
pks_fks as (
  -- pk + fk referencing col
  select
    conrelid as resorigtbl,
    unnest(conkey) as resorigcol
  from pg_constraint
  where contype IN ('p', 'f')
  union
  -- fk referenced col
  select
    confrelid,
    unnest(confkey)
  from pg_constraint
  where contype='f'
),
views as (
  select
    c.oid       as view_id,
    n.nspname   as view_schema,
    c.relname   as view_name,
    r.ev_action as view_definition
  from pg_class c
  join pg_namespace n on n.oid = c.relnamespace
  join pg_rewrite r on r.ev_class = c.oid
  where c.relkind in ('v', 'm') and n.nspname = ANY($1 || $2)
),
transform_json as (
  select
    view_id, view_schema, view_name,
    -- the following formatting is without indentation on purpose
    -- to allow simple diffs, with less whitespace noise
    replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      regexp_replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
        view_definition::text,
      -- This conversion to json is heavily optimized for performance.
      -- The general idea is to use as few regexp_replace() calls as possible.
      -- Simple replace() is a lot faster, so we jump through some hoops
      -- to be able to use regexp_replace() only once.
      -- This has been tested against a huge schema with 250+ different views.
      -- The unit tests do NOT reflect all possible inputs. Be careful when changing this!
      -- -----------------------------------------------
      -- pattern           | replacement         | flags
      -- -----------------------------------------------
      -- `,` is not part of the pg_node_tree format, but used in the regex.
      -- This removes all `,` that might be part of column names.
         ','               , ''
      -- The same applies for `{` and `}`, although those are used a lot in pg_node_tree.
      -- We remove the escaped ones, which might be part of column names again.
      ), E'\\{'            , ''
      ), E'\\}'            , ''
      -- The fields we need are formatted as json manually to protect them from the regex.
      ), ' :targetList '   , ',"targetList":'
      ), ' :resno '        , ',"resno":'
      ), ' :resorigtbl '   , ',"resorigtbl":'
      ), ' :resorigcol '   , ',"resorigcol":'
      -- Make the regex also match the node type, e.g. `{QUERY ...`, to remove it in one pass.
      ), '{'               , '{ :'
      -- Protect node lists, which start with `({` or `((` from the greedy regex.
      -- The extra `{` is removed again later.
      ), '(('              , '{(('
      ), '({'              , '{({'
      -- This regex removes all unused fields to avoid the need to format all of them correctly.
      -- This leads to a smaller json result as well.
      -- Removal stops at `,` for used fields (see above) and `}` for the end of the current node.
      -- Nesting can't be parsed correctly with a regex, so we stop at `{` as well and
      -- add an empty key for the followig node.
      ), ' :[^}{,]+'       , ',"":'              , 'g'
      -- For performance, the regex also added those empty keys when hitting a `,` or `}`.
      -- Those are removed next.
      ), ',"":}'           , '}'
      ), ',"":,'           , ','
      -- This reverses the "node list protection" from above.
      ), '{('              , '('
      -- Every key above has been added with a `,` so far. The first key in an object doesn't need it.
      ), '{,'              , '{'
      -- pg_node_tree has `()` around lists, but JSON uses `[]`
      ), '('               , '['
      ), ')'               , ']'
      -- pg_node_tree has ` ` between list items, but JSON uses `,`
      ), ' '             , ','
      -- `<>` in pg_node_tree is the same as `null` in JSON, but due to very poor performance of json_typeof
      -- we need to make this an empty array here to prevent json_array_elements from throwing an error
      -- when the targetList is null.
      ), '<>'              , '[]'
    )::json as view_definition
  from views
),
target_entries as(
  select
    view_id, view_schema, view_name,
    json_array_elements(view_definition->0->'targetList') as entry
  from transform_json
),
results as(
  select
    view_id, view_schema, view_name,
    (entry->>'resno')::int as view_column,
    (entry->>'resorigtbl')::oid as resorigtbl,
    (entry->>'resorigcol')::int as resorigcol
  from target_entries
),
recursion as(
  select r.*
  from results r
  where view_schema = ANY ($1)
  union all
  select
    view.view_id,
    view.view_schema,
    view.view_name,
    view.view_column,
    tab.resorigtbl,
    tab.resorigcol
  from recursion view
  join results tab on view.resorigtbl=tab.view_id and view.resorigcol=tab.view_column
)
select
  sch.nspname as table_schema,
  tbl.relname as table_name,
  col.attname as table_column_name,
  rec.view_schema,
  rec.view_name,
  vcol.attname as view_column_name
from recursion rec
join pg_class tbl on tbl.oid = rec.resorigtbl
join pg_attribute col on col.attrelid = tbl.oid and col.attnum = rec.resorigcol
join pg_attribute vcol on vcol.attrelid = rec.view_id and vcol.attnum = rec.view_column
join pg_namespace sch on sch.oid = tbl.relnamespace
join pks_fks using (resorigtbl, resorigcol)
order by view_schema, view_name, view_column_name;
</pre>
</details>

**Comment by rinshadka** [comment]
@steve-chavez , Thanks. Actually I am having 10 schemas in database and got this error while running the above query with one particular schema. Later I temporarily removed that schema only from postgREST and was able to start the service properly. 

What may be the issue with that schema?

**Comment by wolfgangwalther** [comment]
> @steve-chavez , Thanks. Actually I am having 10 schemas in database and got this error while running the above query with one particular schema. Later I temporarily removed that schema only from postgREST and was able to start the service properly.
> 
> What may be the issue with that schema?

Ah, that's very interesting. Are you sure you are running `standard_conforming_strings=off`? And does the error go away when running with `standard_conforming_strings=on`? Or did you just add to this issue, because the error message is similar?

If it turns out to be unrelated to `standard_conforming_strings`, we might have our first case of "our json parsing breaks with some view special case view definition". This something I expected to happen at some point in https://github.com/PostgREST/postgrest/pull/1632#issuecomment-740629768. In that case, it would be great if you could reduce the problematic schema to the smallest reproducible example (it's one of the view definitions for sure) - or, if possible, could share the schema in question with us privately.

**Comment by wolfgangwalther** [comment]
> Ah, that's very interesting. Are you sure you are running `standard_conforming_strings=off`? And does the error go away when running with `standard_conforming_strings=on`? Or did you just add to this issue, because the error message is similar?

Ah, I see you already answered that above:

> Also experimentally changed the standard_conforming_strings postgreSQL configuration for postgREST database user, but failed.

Let's track this in a new issue, as it's very likely to be independent of this.

**Comment by steve-chavez** [comment]
> Interestingly, the code for pgFmtLit in SqlFragment.hs seems to use E'' strings, but pgFmtArrayLit does not although I don't know Haskell so could be reading it incorrectly.

> I don't think pgFmtArrayLit is a problem here, because the escaped string will be passed as a value in a parametrized query - not as a string as part of the query itself. Therefore standard_conforming_strings=off doesn't affect this part.

@dgtrapeze Just wanted to reaffirm this in case there's any doubt. The `pgFmtArrayLit` function doesn't produce a [pg string constant](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-STRINGS) but a value that will be parametrized, so there's no need to use the [C-style escape](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-STRINGS).

Also when testing with `standard_conforming_strings = off` on https://github.com/PostgREST/postgrest/pull/1995 the `in` filter didn't produce an error.

Btw, I'm removing the `pgFmtLit` function in https://github.com/PostgREST/postgrest/pull/2027 to make sure we don't mess with escaping string constants anymore.

**Comment by dgtrapeze** [comment]
Great.  I'm using v8.0.0.20211102 without issue (although we got caught out by the change in #1849 which changed the response structure from array to single object when we updated from V7 to V8), but haven't noticed any other issues due to  standard_conforming_strings
